src/frequenz/sdk/actor/_channel_registry.py

```diff
@@ -0,0 +1,151 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""A class that would dynamically create, own and provide access to channels."""
+
+import dataclasses
+import logging
+import traceback
+from typing import TypeVar, cast
+
+from frequenz.channels import Broadcast
+
+_T = TypeVar("_T")
+_logger = logging.getLogger(__name__)
+
+
+class ChannelRegistry:
+    """Dynamically creates, own and provide access to broadcast channels.
+
+    It can be used by actors to dynamically establish a communication channel
+    between each other.
+
+    The registry is responsible for creating channels when they are first requested via
+    the [`get_or_create()`][frequenz.sdk.actor.ChannelRegistry.get_or_create] method.
+
+    The registry also stores type information to make sure that the same channel is not
+    used for different message types.
+
+    Since the registry owns the channels, it is also responsible for closing them when
+    they are no longer needed. There is no way to remove a channel without closing it.
+
+    Note:
+        This registry stores [`Broadcast`][frequenz.channels.Broadcast] channels.
+    """
+
+    def __init__(self, *, name: str) -> None:
+        """Initialize this registry.
+
+        Args:
+            name: A name to identify the registry in the logs. This name is also used as
+                a prefix for the channel names.
+        """
+        self._name = name
+        self._channels: dict[str, _Entry] = {}
+
+    @property
+    def name(self) -> str:
+        """The name of this registry."""
+        return self._name
+
+    def message_type(self, key: str) -> type:
+        """Get the message type of the channel for the given key.
+
+        Args:
+            key: The key to identify the channel.
+
+        Returns:
+            The message type of the channel.
+
+        Raises:
+            KeyError: If the channel does not exist.
+        """
+        entry = self._channels.get(key)
+        if entry is None:
+            raise KeyError(f"No channel for key {key!r} exists.")
+        return entry.message_type
+
+    def __contains__(self, key: str) -> bool:
+        """Check whether the channel for the given `key` exists."""
+        return key in self._channels
+
+    def get_or_create(
+        self, message_type: type[_T], key: str, *, check_type: bool = True
```

Comment:
I have the feeling that the argument `check_type` is redundant in the sense that it should always be true. Or I least I don't have any scenario in mind where the user needs to set `check_type` to false providing always a `message_type` as it is mandatory.

In case we have a scenario where the `check_type` should be skipped, then I think a `get(self, key: str)`  method can be provided but without encouraging the user to use it. Probably not  a very good idea

Reply:
I agree the API is weird.

I added `check_type` as a escape mechanism because it is not possible (or not easy) to check for sub-types, so if you have `class A: pass; class B(A): pass` and then do `get_or_create(B, k); get_or_create(A, k)` the second call will fail even when it is safe to do so.

It is still probably a very niche use case, so I'd be fine to removing that option until we really need it (I thought I might needed it somewhere, but I think it is not really used).